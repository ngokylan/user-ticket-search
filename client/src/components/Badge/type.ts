import { ReactNode } from 'react';

export type BadgeType =
  | 'grey'
  | 'outlined'
  | 'warning'
  | 'info'
  | 'success'
  | 'danger';
export const BadgeTypes: { [key: string]: BadgeType } = {
  grey: 'grey',
  outlined: 'outlined',
  warning: 'warning',
  info: 'info',
  success: 'success',
  danger: 'danger',
};

export type BadgeRole = 'info' | 'status' | 'label';

export type BadgeProps = {
  /**
   * Badge type, use one from the following:
   * BadgeType = 'grey' | 'outlined' | 'warning' | 'info' | 'success' | 'danger'
   */
  type: BadgeType;
  /**
   * Badge role, use one from the following:
   * BadgeRole = 'info' | 'status' | 'label'
   */
  role?: BadgeRole;
  /** Custom id attribute */
  id?: string;
  /** Add custom class to class attribute */
  className?: string;
  /** To enable popover Icon */
  hasPopoverIcon?: boolean;
  /** Badge icon */
  icon?: ReactNode;
  /** The content of a badge */
  children?: ReactNode;
  /** To render badge in circular */
  isCircular?: boolean;
};
