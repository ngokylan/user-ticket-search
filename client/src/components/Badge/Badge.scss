@import '../../_lib/style/index';

.badge {
  $this: &;
  display: inline-flex;
  align-items: center;
  align-content: center;
  justify-content: center;
  @include text-style-body-xs;
  letter-spacing: 1.25px;
  font-weight: 500;
  text-transform: uppercase;
  background-color: $gray-700;
  color: $white;
  height: 24px;
  line-height: 24px;
  padding-left: 4px;
  padding-right: 4px;
  border-radius: 12px;
  min-width: 24px;
  text-align: center;
  letter-spacing: 0.25px;
  max-width: 100%;
  vertical-align: middle;

  .icon svg {
    font-size: 16px;
  }

  &__icon {
    display: inline-flex;
    align-items: center;

    .icon {
      display: inline-flex;
      align-items: center;
    }
  }

  &__label {
    display: inline-block;
    padding-left: 4px;
    padding-right: 4px;
    overflow: hidden;
    text-overflow: ellipsis;
    white-space: nowrap;
  }

  &__popover-icon {
    opacity: 0.67;
  }

  &--is-circular {
    width: 24px;
    letter-spacing: 0;
    padding: 0;

    .badge__label {
      white-space: nowrap;
      overflow: hidden;
      padding: 0;
      text-align: center;
    }
  }

  &--grey {
    background-color: $gray-700;
    color: $white;

    &#{$this}--lighter {
      color: $gray-700;
      background-color: $gray-200;
    }
  }

  &--info {
    background-color: $info;

    &#{$this}--lighter {
      color: $gray-700;
      background-color: $info-bg;
    }
  }

  &--success {
    background-color: $success;

    &#{$this}--lighter {
      color: $gray-700;
      background-color: $success-bg;
    }
  }

  &--warning {
    background-color: $warning;

    &#{$this}--lighter {
      color: $gray-700;
      background-color: $warning-bg;
    }

    color: $gray-800;
  }

  &--danger {
    background-color: $danger;

    &#{$this}--lighter {
      color: $gray-700;
      background-color: $danger-bg;
    }
  }

  &--outlined {
    background-color: #fff;
    border: 1px solid $gray-700;
    color: $gray-800;
  }

  &--role-status {
  }

  &--role-label {
    height: 20px;
    line-height: 18px;
    border-radius: 3px;
    border-color: $gray-700;
    padding-left: 3px;
    padding-right: 3px;

    #{$this}__label {
      letter-spacing: 0.4px;
      font-size: 12px;
      color: $gray-700;
      line-height: 18px;
      padding: 0;
    }
  }

  &--role-info {
    background-color: $black;
    border-radius: 3px;
    text-transform: none;
    padding: 15px 5px 15px 5px;
    font-weight: normal;

    .badge__label {
      padding-left: 5px;
      padding-right: 0;
    }

    .icon svg {
      font-size: 20px;
    }
  }
}
