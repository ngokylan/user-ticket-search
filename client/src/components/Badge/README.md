Badge 

### When To Use
A small count and labeling component. May be used in tables to show important information. Generally not clickable.

### Examples
Warning badge
```js
<Badge type={'warning'}>PENDING</Badge>
```

Success badge
```js
<Badge type={'success'}>FINALISED</Badge>
```

Info badge
```js
<Badge type={'info'}>INFO</Badge>
```

Danger badge
```js
<Badge type={'danger'}>DANGER</Badge>
```

Grey badge
```js
<Badge>999</Badge>
```

Outlined badge
```js
<Badge type={'outlined'}>999</Badge>
```
