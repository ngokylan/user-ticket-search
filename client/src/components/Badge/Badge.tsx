import React from 'react';
import { getClass } from '../../_lib/helper';
import { BadgeProps } from './type';
import './Badge.scss';

function Badge({
  id,
  icon,
  role,
  type,
  children,
  className,
  hasPopoverIcon,
  isCircular,
}: BadgeProps) {
  return (
    <span
      id={id}
      className={getClass('badge', className, {
        [type]: type,
        lighter: children === '0' || children === 0,
        'is-circular': isCircular,
        [`role-${role}`]: role,
      })}
      data-testid={`badge-${id || 'default'}`}
    >
      {icon && <span className="badge__icon">{icon}</span>}
      {children && <span className="badge__label">{children}</span>}      
    </span>
  );
}

Badge.defaultProps = {
  role: 'status',
  type: 'grey',
};

export default Badge;
