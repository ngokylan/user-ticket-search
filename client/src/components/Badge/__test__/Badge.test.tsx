import React from 'react';
import { mount } from 'enzyme';
import Badge, { BadgeProps } from '..';

const defaultProps: BadgeProps = {
  role: 'status',
  type: 'grey',
};

const mountComponent = (customProps: any = {}) => {
  const props = { ...defaultProps, ...customProps };
  const wrapper = mount(<Badge {...props} />);
  const component = wrapper.find('.badge');

  return { props, wrapper, component };
};

describe('Badge', () => {
  describe('rendering', () => {
    it('should render component without any crash', () => {
      const { wrapper } = mountComponent();
      expect(wrapper).toHaveLength(1);
    });

    it('should be match with snapshot', () => {
      const { wrapper } = mountComponent();
      expect(wrapper).toMatchSnapshot();
    });
  });

  describe('props', () => {
    describe('default', () => {
      it('should have expected default props', () => {
        expect(JSON.stringify(defaultProps)).toBe(
          JSON.stringify(Badge.defaultProps)
        );
      });

      it.each([
        ['badge-default', {}],
        ['badge-custom-id', { id: 'custom-id' }],
      ])('should render test ID properly', (expectedTestId, props) => {
        const { component } = mountComponent(props);
        expect(component.prop('data-testid')).toBe(expectedTestId);
      });
    });

    describe('children', () => {
      it('should render children', () => {
        const { wrapper } = mountComponent({ children: 'Hello world' });
        expect(wrapper.find('.badge__label').text()).toBe('Hello world');
      });

      it('should not render children', () => {
        const { wrapper } = mountComponent();
        expect(wrapper.find('.badge__label')).toHaveLength(0);
      });
    });

    describe('custom', () => {
      it.each([['id', 'test', { id: 'test' }]])(
        'should render prop %s as %s',
        (name, expected, props) => {
          const { component } = mountComponent(props);
          expect(component.prop(name)).toBe(expected);
        }
      );

      it.each([
        ['grey', '.badge--grey'],
        ['outlined', '.badge--outlined'],
        ['warning', '.badge--warning'],
        ['info', '.badge--info'],
        ['success', '.badge--success'],
        ['danger', '.badge--danger'],
      ])(
        'should render proper class name for prop type as %s',
        (type, query) => {
          const { wrapper } = mountComponent({ type });
          expect(wrapper.find(query)).toHaveLength(1);
        }
      );
    });
  });

  describe('interaction', () => {
    it.each([
      ['should', '.badge--lighter', 1, { children: 0 }],
      ['should', '.badge--lighter', 1, { children: '0' }],
      ['should not', '.badge--lighter', 0, { children: undefined }],
      ['should not', '.badge--lighter', 0, { children: null }],
      [
        'should not',
        '.badge--lighter',
        0,
        { children: 'some other text' },
      ],
      ['should not', '.badge--lighter', 0, { children: 10 }],
      ['should', '.badge__icon', 1, { icon: 'icon' }],
      ['should not', '.badge__icon', 0, {}],      
    ])('%s have %s', (name, element, expected, props) => {
      const { wrapper } = mountComponent(props);
      expect(wrapper.find(element)).toHaveLength(expected);
    });
  });
});
