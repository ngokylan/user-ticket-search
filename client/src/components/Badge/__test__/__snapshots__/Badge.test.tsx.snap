// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Badge rendering should be match with snapshot 1`] = `
<span
  className="badge badge--grey badge--role-status"
  data-testid="badge-default"
/>
`;
